File: android/beagle/src/main/java/br/com/zup/beagle/android/action/Action.kt
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package br.com.zup.beagle.android.action

import br.com.zup.beagle.android.context.Bind
import br.com.zup.beagle.android.utils.Observer
import br.com.zup.beagle.android.utils.getWithCaller
import br.com.zup.beagle.android.widget.RootView

interface Action {
    fun execute(rootView: RootView)

    /**
     * Retrieves the value if its not an expression or observes the changes if its an expression
     * @property rootView from buildView
     * @property bind is the value that will retrieved or observed
     * @property observes is function that will be called when a expression is evaluated
     */
    fun <T> evaluateBinding(
        rootView: RootView,
        bind: Bind<T>,
        observes: Observer<T>? = null
    ): T? {
        return bind.getWithCaller(rootView, this, observes)
    }
}

File: iOS/Sources/Beagle/Sources/Action/Action.swift
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import UIKit

/// Markup to define an action to be triggered in response to some event
public protocol Action: Decodable {

    /// can be used to override or extend the global `AnalyticsConfig`.
    ///
    /// - when `nil`: analytics behavior for this action will be determined by global `AnalyticsConfig`.
    var analytics: ActionAnalyticsConfig? { get }
    
    func execute(controller: BeagleController, origin: UIView)
}

public protocol AsyncAction: Action {
    var onFinish: [Action]? { get set }
}

/// Defines a representation of an unknown Action
public struct UnknownAction: Action {
    public let type: String
    public var analytics: ActionAnalyticsConfig? { return nil }
    
    public init(type: String) {
        self.type = type
    }
    
    public func execute(controller: BeagleController, origin: UIView) {
        controller.dependencies.logger.log(Log.decode(.decodingError(type: "error trying to execute unknown action")))
    }
}

extension Action {
    public func closureToRetrySameAction(controller: BeagleController, origin: UIView) -> BeagleRetry {
        return { [weak controller] in
            guard let controller = controller else { return }
            self.execute(controller: controller, origin: origin)
        }
    }
}
